Customize traceback for calls to symbolically-traced code

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51648 Customize traceback for calls to symbolically-traced code**

The following code will throw during the call to `traced(5)`:
```python
class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced = fx.symbolic_trace(M())
traced(5)
```

Traceback before:
```
Traceback (most recent call last):
  File "test/tinytest.py", line 26, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

Traceback after:
```
Traceback (most recent call last):
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 334, in wrapped_call
    return torch.nn.Module.__call__(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_1>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int

Call using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    relu_1 = dot_1.relu();  dot_1 = None

    return relu_1

Traceback (most recent call last):
  File "test/tinytest.py", line 21, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 342, in wrapped_call
    raise e.with_traceback(None)
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

(Note that the same `TypeError` is thrown despite modifying the traceback.)

Differential Revision: [D26424005](https://our.internmc.facebook.com/intern/diff/D26424005)

Commit:
Customize traceback for calls to symbolically-traced code

[ghstack-poisoned]

Commit:
Update on "Customize traceback for calls to symbolically-traced code"




[ghstack-poisoned]

jamesr66a commented on test/test_fx.py:
```diff
@@ -1597,6 +1597,42 @@ def test_no_mutation(self):
+
+        traced(5)
+
+        with patch("sys.stderr", new=StringIO()) as fake_err:
```
`TestFX` inherits from `JitTestCase`, which defines a `capture_stdout` context manager: https://github.com/pytorch/pytorch/blob/18a7ec7d7d0a6726e73dc53a6ee1044b8f2513b1/torch/testing/_internal/jit_utils.py#L100

Maybe you can add a similar `capture_stderr` manager on `JitTestCase` and use that here

jamesr66a commented on test/test_fx.py:
```diff
@@ -1597,6 +1597,42 @@ def test_no_mutation(self):
+
+        traced = torch.fx.symbolic_trace(M())
+
+        out = [n for n in list(traced.graph.nodes) if n.op == "output"]
```
nit: Don't think you need to put all the nodes in a list for this for-comp to work

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -296,18 +301,65 @@ def recompile(self) -> None:
+        # symbolically-traced code was run with an invalid input, the
+        # user would see the source of the error as coming from
+        # `File "<eval_with_key_N">`, where N is the number of times
+        # `recompile` had been called on the GraphModule. We use this
```
`N` is actually the result of a global counter, so only monotonic with how many times `recompile` was called on this specific GraphModule but not 1:1

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -296,18 +301,65 @@ def recompile(self) -> None:
+        # return the traceback itself, a message explaining that the
+        # error occurred in a traced Module's generated forward
+        # function, and five lines of context surrounding the faulty
+        # faulty line
```
nit: `faulty` twice

Commit:
Update on "Customize traceback for calls to symbolically-traced code"


The following code will throw during the call to `traced(5)`:
```python
class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced = fx.symbolic_trace(M())
traced(5)
```

Traceback before:
```
Traceback (most recent call last):
  File "test/tinytest.py", line 26, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

Traceback after:
```
Traceback (most recent call last):
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 349, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_41>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
Call to “exec” using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None
    return dot_1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE    relu_1 = [output].relu();  output = None
```

(Note that the same `TypeError` is thrown despite modifying the traceback.)


[ghstack-poisoned]

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -296,18 +301,65 @@ def recompile(self) -> None:
+        # we use the instance attribute `_cls_call` to store the
+        # original implementation of `__call__`
+        if not self._cls_call:
+            self._cls_call = cls.__call__
```
Can we just call `torch.nn.Module.__call__` instead of storying this away?

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -296,18 +301,65 @@ def recompile(self) -> None:
+        # faulty line
+        def generate_error_message(tb: TracebackType) -> str:
+            # auxiliary variables (for readability)
+            frame_summary = traceback.extract_tb(tb)[-1]
```
This is implicitly assuming that the error is occurring in the `GraphModule` code itself, but in reality an exception propagating through here could come from any transitive call. For example:
```
import torch
import torch.fx

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(x)

f = torch.fx.symbolic_trace(Foo())
f(torch.rand(5, 5))
"""
Traceback (most recent call last):
  File "/data/users/jamesreed/pytorch/torch/fx/graph_module.py", line 350, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/data/users/jamesreed/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 3, in forward
    linear_1 = self.linear(x);  x = None
  File "/data/users/jamesreed/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/data/users/jamesreed/pytorch/torch/nn/modules/linear.py", line 94, in forward
    return F.linear(input, self.weight, self.bias)
  File "/data/users/jamesreed/pytorch/torch/nn/functional.py", line 1755, in linear
    ret = torch.addmm(bias, input, weight.t())
RuntimeError: mat1 and mat2 shapes cannot be multiplied (5x5 and 3x4)
Call to “exec” using an FX-traced Module, line 1755 of the traced Module’s generated forward function:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
Traceback (most recent call last):
  File "relu_test.py", line 13, in <module>
    f(torch.rand(5, 5))
  File "/data/users/jamesreed/pytorch/torch/fx/graph_module.py", line 358, in wrapped_call
    raise e.with_traceback(None)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (5x5 and 3x4)
"""
```

In this case the extra context is actually incorrect and probably confusing. I think maybe at 351 you could peel off the top frame from the traceback and check if that filename contains `eval_with_cache`, then dispatch into `generate_error_message`, otherwise just rethrow

jamesr66a replied:
Can we pass in `topmost_framesummary` here rather than doing `extract_tb` again?

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -296,18 +301,65 @@ def recompile(self) -> None:
+            frame_summary = traceback.extract_tb(tb)[-1]
+            err_lineno = frame_summary.lineno
+            err_line_len = len(frame_summary.line)
+            all_src_lines = list(_eval_cache.values())[-1]
```
Pulling off the back of the eval cache is not the right thing to do here. In python, global variable references are looked up on every call and not closed over. This creates an incorrect context:

```
import torch
import torch.fx

class M(torch.nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced_dot = torch.fx.symbolic_trace(M())

class N(torch.nn.Module):
    def forward(self, x):
        return torch.relu(x)

traced_relu = torch.fx.symbolic_trace(N())

traced_dot(5)
"""
Traceback (most recent call last):
  File "/data/users/jamesreed/pytorch/torch/fx/graph_module.py", line 350, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/data/users/jamesreed/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
Call to “exec” using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    relu_1 = torch.relu(x);  x = None
    return relu_1
    
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
Traceback (most recent call last):
  File "eval_cache.py", line 20, in <module>
    traced_dot(5)
  File "/data/users/jamesreed/pytorch/torch/fx/graph_module.py", line 358, in wrapped_call
    raise e.with_traceback(None)
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
"""

I think a simple thing to do here instead is `all_src_lines = _eval_cache[frame_summary.filename]`
```

jamesr66a reviewed (commented):
Cool! 

I'm not sure if I really understand the traceback format:
```
Call to “exec” using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None
    return dot_1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE    relu_1 = [output].relu();  output = None
```

In the TorchScript error context the tilde context would appear directly after the line where the error occurred and its width would be constrained to the width of the expression that has the error. I think it's pretty confusing that it appears after everything and covers the whole line

Also missing a newline after `HERE`

jamesr66a commented on test/test_fx.py:
```diff
@@ -1597,6 +1597,40 @@ def test_no_mutation(self):
+        traced = torch.fx.symbolic_trace(M())
+
+        out = [n for n in list(traced.graph.nodes) if n.op == "output"]
+        relu_out = traced.graph.call_method(method_name='relu', args=(out,))
```
Couple things:

1. `out` is List[Node] here
2. The value of an `output` node is not meaningful, what you probably wanted was `out.args[0]`

I think this section should be:
```
        out_nodes = [n for n in list(traced.graph.nodes) if n.op == "output"]
        out = out_nodes[-1]
        relu_out = traced.graph.call_method(method_name='relu', args=(out,))
        out.replace_all_uses_with(relu_out)
```

jamesr66a replied:
I had a typo as well, I think the relu_out line should be:

```
        relu_out = traced.graph.call_method(method_name='relu', args=(out.args[0],))
```

Since we want to take the value the output node references and not the output node itself

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -296,18 +301,65 @@ def recompile(self) -> None:
+
+            # constiuent substrings of the error message
+            tb_repr = traceback.format_exc()
+            custom_msg = ("Call to “exec” using an FX-traced Module, "
```
BTW saying call to "exec" is confusing, since exec is the function that take source code and emits a code object. It would be more accurate to say call to the code object, but that's still too far into the weeds of the implementation details IMO. I think this should just say something like "Call to generated forward..."

Commit:
Update on "Customize traceback for calls to symbolically-traced code"


The following code will throw during the call to `traced(5)`:
```python
class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced = fx.symbolic_trace(M())
traced(5)
```

Traceback before:
```
Traceback (most recent call last):
  File "test/tinytest.py", line 26, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

Traceback after:
```
Traceback (most recent call last):
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 336, in wrapped_call
    return torch.nn.Module.__call__(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_50>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int

Call using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    return dot_1

    relu_1 = relu_1.relu();  relu_1 = None
```

(Note that the same `TypeError` is thrown despite modifying the traceback.)


[ghstack-poisoned]

Commit:
Update on "Customize traceback for calls to symbolically-traced code"


The following code will throw during the call to `traced(5)`:
```python
class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced = fx.symbolic_trace(M())
traced(5)
```

Traceback before:
```
Traceback (most recent call last):
  File "test/tinytest.py", line 26, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

Traceback after:
```
Traceback (most recent call last):
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 336, in wrapped_call
    return torch.nn.Module.__call__(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_50>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int

Call using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    return dot_1

    relu_1 = relu_1.relu();  relu_1 = None
```

(Note that the same `TypeError` is thrown despite modifying the traceback.)


[ghstack-poisoned]

jamesr66a commented on test/test_fx.py:
```diff
@@ -1597,6 +1597,31 @@ def test_no_mutation(self):
+
+        self.assertIn("Call using an FX-traced Module, line 4 of the "
+                      "traced Module’s generated forward function:",
+                      captured[0])
```
I think it'd also be a good idea to add tests ensuring that this is not firing for exceptions thrown within submodules, like the example I put in a comment before

Commit:
Update on "Customize traceback for calls to symbolically-traced code"


The following code will throw during the call to `traced(5)`:
```python
class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced = fx.symbolic_trace(M())
traced(5)
```

Traceback before:
```
Traceback (most recent call last):
  File "test/tinytest.py", line 26, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

Traceback after:
```
Traceback (most recent call last):
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 336, in wrapped_call
    return torch.nn.Module.__call__(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_50>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int

Call using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    return dot_1

    relu_1 = relu_1.relu();  relu_1 = None
```

(Note that the same `TypeError` is thrown despite modifying the traceback.)


[ghstack-poisoned]

jamesr66a commented on test/test_fx.py:
```diff
@@ -1597,6 +1598,62 @@ def test_no_mutation(self):
+        # manager doesn't work in this case even though it works on the
+        # other `test_custom_traceback` test. I'd put the "woman
+        # shrugging" emoji in here but I don't want to press my luck
+        # in terms of things unexpectedly failing
```
I think it's because when you intercept the exception propagation and print out the GraphModule context, that happens in a a callee frame with respect to this test, but when the exception isn't intercepted it propagates out through here, exits the test, and prints to `stderr` in some test harness function (caller of this function)

jamesr66a commented on test/test_fx.py:
```diff
@@ -1597,6 +1598,62 @@ def test_no_mutation(self):
+        out_nodes = [n for n in traced.graph.nodes if n.op == "output"]
+        out = out_nodes[-1]
+        relu_out = traced.graph.call_method(method_name='relu', args=(out.args[0],))
+        out.replace_all_uses_with(relu_out)
```
Hate to keep commenting on this test case, but two things:
1) I think you want to insert the `call_method` call before the output node
2) Similarly to args above, I think you should do `out.args[0].replace_all_uses_with(relu_out)`

jamesr66a reviewed (approved):
Nice. As discussed, let's wait until the 1.8 branch cut happens before merging this.

BTW is the error printout in the description up to date?

Commit:
Update on "Customize traceback for calls to symbolically-traced code"


The following code will throw during the call to `traced(5)`:
```python
class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced = fx.symbolic_trace(M())
traced(5)
```

Traceback before:
```
Traceback (most recent call last):
  File "test/tinytest.py", line 26, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

Traceback after:
```
Traceback (most recent call last):
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return torch.nn.Module.__call__(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_1>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int

Call using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    relu_1 = dot_1.relu();  dot_1 = None

    return relu_1
```

(Note that the same `TypeError` is thrown despite modifying the traceback.)


[ghstack-poisoned]

Commit:
Update on "Customize traceback for calls to symbolically-traced code"


The following code will throw during the call to `traced(5)`:
```python
class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced = fx.symbolic_trace(M())
traced(5)
```

Traceback before:
```
Traceback (most recent call last):
  File "test/tinytest.py", line 26, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

Traceback after:
```
Traceback (most recent call last):
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return torch.nn.Module.__call__(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_1>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int

Call using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    relu_1 = dot_1.relu();  dot_1 = None

    return relu_1
```

(Note that the same `TypeError` is thrown despite modifying the traceback.)


[ghstack-poisoned]

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -296,18 +297,50 @@ def recompile(self) -> None:
-
         def print_full_traceback(exctype, value, tb):
+            if self.foo:
+                return
```
?

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -296,18 +297,50 @@ def recompile(self) -> None:
             traceback.print_exception(exctype, value, tb)
 
+        # Previously, if an error occurred when valid
+        # symbolically-traced code was run with an invalid input, pytes the
```
What is `pytes`

Commit:
Update on "Customize traceback for calls to symbolically-traced code"


The following code will throw during the call to `traced(5)`:
```python
class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced = fx.symbolic_trace(M())
traced(5)
```

Traceback before:
```
Traceback (most recent call last):
  File "test/tinytest.py", line 26, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

Traceback after:
```
Traceback (most recent call last):
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 334, in wrapped_call
    return torch.nn.Module.__call__(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_1>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int

Call using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    relu_1 = dot_1.relu();  dot_1 = None

    return relu_1

Traceback (most recent call last):
  File "test/tinytest.py", line 21, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 342, in wrapped_call
    raise e.with_traceback(None)
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

(Note that the same `TypeError` is thrown despite modifying the traceback.)

Differential Revision: [D26424005](https://our.internmc.facebook.com/intern/diff/D26424005)

[ghstack-poisoned]

Commit:
Update on "Customize traceback for calls to symbolically-traced code"


The following code will throw during the call to `traced(5)`:
```python
class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced = fx.symbolic_trace(M())
traced(5)
```

Traceback before:
```
Traceback (most recent call last):
  File "test/tinytest.py", line 26, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 338, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

Traceback after:
```
Traceback (most recent call last):
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 334, in wrapped_call
    return torch.nn.Module.__call__(self, *args, **kwargs)
  File "/home/ansley/local/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_1>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int

Call using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    w = self.W
    dot_1 = torch.dot(w, x);  w = x = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    relu_1 = dot_1.relu();  dot_1 = None

    return relu_1

Traceback (most recent call last):
  File "test/tinytest.py", line 21, in <module>
    traced(5)
  File "/home/ansley/local/pytorch/torch/fx/graph_module.py", line 342, in wrapped_call
    raise e.with_traceback(None)
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

(Note that the same `TypeError` is thrown despite modifying the traceback.)

Differential Revision: [D26424005](https://our.internmc.facebook.com/intern/diff/D26424005)

[ghstack-poisoned]
